py/dml/codegen.py

```diff
@@ -1573,10 +1573,13 @@ def stmt_log(stmt, location, scope):
         if location.method():
             key = mkLit(site, '(uint64)NULL', TInt(64, False))
```

Comment:
This can be modified to instead be based on the indices provided to the method call, which will allow objects of an array to be independent of one another. You can leverage `encode_indices` from ctree for this purpose.

Reply:
I think my current implementation just ends up using _identity_to_key again, just with an ObjIdentity instead.

Reply:
I'm slightly skeptical about using full-fledged identities in the non-shared case -- I feel like that would cause the outermost hash table to become unnecessarily big.  Using `ObjIdentity` is certainly easier, but I feel only basing the key on indices (in this case) would be somewhat more morally correct.

Reply:
I vote for ObjIdentity if it makes code cleaner; we don't expect this to be anywhere close to expensive in practice. Showing a message where a human can see it costs magnitudes more than a few bytes of memory, so as long as table is only populated when messages are logged, the relative cost is negligible.